subsys/settings/src/settings_nvs.c

```diff
@@ -0,0 +1,238 @@
+		}
+
+		if ((rc1 <= 0) || (rc2 <= 0)) {
+			/* NVS is dirty, clean it */
+			if (name_id == cf->last_name_id) {
+				cf->last_name_id--;
+				nvs_write(&cf->cf_nvs, NVS_NAMECNT_ID,
+					  &cf->last_name_id, sizeof(u16_t));
+			}
+			nvs_delete(&cf->cf_nvs, name_id);
+			nvs_delete(&cf->cf_nvs, name_id + NVS_NAME_ID_OFFSET);
+			continue;
+		}
+
+		/* Found a name, this might not include a trailing \0 */
+		name[rc1] = '\0';
+
+		if (subtree && !settings_name_steq(name, subtree, NULL)) {
+			continue;
+		}
+
+		ch = settings_parse_and_lookup(name, &name_argv);
+		if (!ch) {
+			continue;
+		}
+
+		read_fn_arg.fs = &cf->cf_nvs;
+		read_fn_arg.id = name_id + NVS_NAME_ID_OFFSET;
+		ch->h_set(name_argv, rc2, settings_nvs_read_fn,
+			  (void *) &read_fn_arg);
+	}
+	return 0;
+}
+
+/* ::csi_save implementation */
+static int settings_nvs_save(struct settings_store *cs, const char *name,
+			     const char *value, size_t val_len)
+{
+	struct settings_nvs *cf = (struct settings_nvs *)cs;
+	char rdname[SETTINGS_MAX_NAME_LEN + SETTINGS_EXTRA_LEN + 1];
+	u16_t name_id, write_name_id;
+	bool delete, write_name;
+	int rc = 0;
+
+	if (!name) {
+		return -EINVAL;
+	}
+
+	/* Find out if we are doing a delete */
+	delete = ((value == NULL) || (val_len == 0));
+
+	name_id = cf->last_name_id + 1;
+	write_name_id = cf->last_name_id + 1;
+	write_name = true;
+
+	while (1) {
+		name_id--;
+		if (name_id == NVS_NAMECNT_ID) {
+			break;
+		}
+
+		rc = nvs_read(&cf->cf_nvs, name_id, &rdname, sizeof(rdname));
+
+		if (rc < 0) {
+			/* Error or entry not found */
+			if (rc == -ENOENT) {
+				write_name_id = name_id;
+			}
+			continue;
+		}
+
+		rdname[rc] = '\0';
+
+		if (strcmp(name, rdname)) {
+			continue;
+		}
+
+		if ((delete) && (name_id == cf->last_name_id)) {
+			cf->last_name_id--;
+			rc = nvs_write(&cf->cf_nvs, NVS_NAMECNT_ID,
+				       &cf->last_name_id, sizeof(u16_t));
+		}
+
+		if (delete) {
+			rc = nvs_delete(&cf->cf_nvs, name_id);
+			rc = nvs_delete(&cf->cf_nvs, name_id +
+					NVS_NAME_ID_OFFSET);
+
+			return 0;
+		}
+		write_name_id = name_id;
+		write_name = false;
+		break;
+	}
+
+	if (delete) {
+		return -ENOENT;
+	}
+
+	/* No free IDs left. */
+	if (write_name_id == NVS_NAMECNT_ID + NVS_NAME_ID_OFFSET) {
+		return -ENOMEM;
+	}
+
+	/* write the value */
+	rc = nvs_write(&cf->cf_nvs, write_name_id + NVS_NAME_ID_OFFSET,
+		       value, val_len);
+
+	/* write the name if required */
+	if (write_name) {
+		rc = nvs_write(&cf->cf_nvs, write_name_id, name, strlen(name));
+	}
+
+	/* update the last_name_id and write to flash if required*/
+	if (write_name_id > cf->last_name_id) {
+		cf->last_name_id = write_name_id;
+		rc = nvs_write(&cf->cf_nvs, NVS_NAMECNT_ID, &cf->last_name_id,
+			       sizeof(u16_t));
+	}
+
+	if (rc < 0) {
+		return rc;
+	}
+
+	return 0;
+}
+
+/* Initialize the nvs backend. */
+int settings_nvs_backend_init(struct settings_nvs *cf)
+{
+	int rc;
+	u16_t last_name_id;
+
+	rc = nvs_init(&cf->cf_nvs, DT_FLASH_DEV_NAME);
```

Comment:
The use of `DT_FLASH_DEV_NAME` limits the nvs file system to be on the main flash device, maybe it is better to pass this the `DT_FLASH_DEVICE` as a const char to `settings_nvs_backend_init()` or to extend settings_nvs with a `const char *flash_device_name`